File: Backend/app/schemas/candidate.py
# backend/app/schemas/candidate.py

from pydantic import BaseModel, UUID4
from typing import Optional
from datetime import datetime


# ===========================
# Schemas for RankedCandidate
# ===========================

class RankedCandidateBase(BaseModel):
    """Base schema for a ranked candidate derived from search results."""
    user_id: UUID4
    jd_id: UUID4
    profile_id: UUID4
    rank: Optional[int] = None
    match_score: Optional[float] = None
    strengths: Optional[str] = None
    favorite: bool = False  # Changed to non-optional with default False
    save_for_future: bool = False
    send_to_recruiter: Optional[UUID4] = None
    outreached: bool = False
    linkedin_url: Optional[str] = None

    # --- NEW FIELDS ---
    contacted: bool = False
    stage: str = "In Consideration"
    # --- END NEW FIELDS ---


class RankedCandidateCreate(RankedCandidateBase):
    """Schema for creating a ranked candidate entry."""
    pass


class RankedCandidateUpdate(BaseModel):
    """Schema for updating fields of an existing ranked candidate."""
    rank: Optional[int] = None
    match_score: Optional[float] = None
    strengths: Optional[str] = None
    favorite: Optional[bool] = None
    save_for_future: Optional[bool] = None
    send_to_recruiter: Optional[UUID4] = None
    outreached: Optional[bool] = None
    linkedin_url: Optional[str] = None

    # --- NEW FIELDS ---
    contacted: Optional[bool] = None
    stage: Optional[str] = None
    # --- END NEW FIELDS ---


class RankedCandidate(RankedCandidateBase):
    """Response schema with full ranked candidate details."""
    rank_id: UUID4
    created_at: datetime

    class Config:
        from_attributes = True


# =====================================
# Schemas for RankedCandidateFromResume
# =====================================

class RankedCandidateFromResumeBase(BaseModel):
    """Base schema for ranked candidates derived from uploaded resumes."""
    user_id: UUID4
    jd_id: UUID4
    resume_id: UUID4
    rank: Optional[int] = None
    match_score: Optional[float] = None
    strengths: Optional[str] = None
    favorite: bool = False  # Changed to non-optional with default False
    save_for_future: bool = False
    send_to_recruiter: Optional[UUID4] = None
    outreached: bool = False
    linkedin_url: Optional[str] = None

    # --- NEW FIELDS ---
    contacted: bool = False
    stage: str = "In Consideration"
    # --- END NEW FIELDS ---


class RankedCandidateFromResumeCreate(RankedCandidateFromResumeBase):
    """Schema for creating a ranked candidate from a resume."""
    pass


class RankedCandidateFromResumeUpdate(BaseModel):
    """Schema for updating ranked candidate from resume."""
    rank: Optional[int] = None
    match_score: Optional[float] = None
    strengths: Optional[str] = None
    favorite: Optional[bool] = None
    save_for_future: Optional[bool] = None
    send_to_recruiter: Optional[UUID4] = None
    outreached: Optional[bool] = None
    linkedin_url: Optional[str] = None

    # --- NEW FIELDS ---
    contacted: Optional[bool] = None
    stage: Optional[str] = None
    # --- END NEW FIELDS ---


class RankedCandidateFromResume(RankedCandidateFromResumeBase):
    """Response schema with full details of a ranked candidate from resume."""
    rank_id: UUID4
    created_at: datetime

    class Config:
        from_attributes = True

# --- NEW SCHEMAS FOR UPDATES ---
class CandidateStageUpdate(BaseModel):
    stage: str

class CandidateContactedUpdate(BaseModel):
    contacted: bool
# --- END NEW SCHEMAS ---

File: Frontend/src/types/candidate.ts
// Frontend/src/types/candidate.ts

// Defines the possible statuses used in some UI bits.
export type CandidateStatus = 'Favourited' | 'Contacted';

// Define the canonical stages used in the Pipeline UI.
// Make sure these strings match what's used on the backend (server_default = "In Consideration").
export type CandidateStage = 'In Consideration' | 'Interviewing' | 'Offer Extended' | 'Rejected';

// An array of stages, used to populate dropdowns on the Pipeline page.
export const candidateStages: CandidateStage[] = [
  'In Consideration',
  'Interviewing',
  'Offer Extended',
  'Rejected',
];

/**
 * This interface is specifically for the Pipeline page, which has a different
 * data structure and UI requirements than the search results.
 */
export interface PipelineCandidate {
  id: string;
  name: string;
  role: string;
  company: string;
  status: CandidateStatus;
  stage: CandidateStage;
}

/**
 * Primary Candidate interface used across Search and Pipeline pages.
 * Fields returned from the backend API should map to this shape.
 *
 * Note:
 * - Many fields are optional because candidates can originate from different
 * sources (web 'search' table vs resume table).
 * - `source` indicates which ranked table the row came from.
 */
export interface Candidate {
  // Unique identifier for the ranked row (ranked_candidates.rank_id or ranked_candidates_from_resume.rank_id)
  rank_id: string;

  // If web/profile sourced, profile_id will be present (maps to `search.profile_id`)
  profile_id?: string;

  // If resume-sourced, resume_id will be present
  resume_id?: string;

  // Which source table this row came from
  source?: 'ranked_candidates' | 'ranked_candidates_from_resume';

  // Matching / ranking metadata
  match_score: number | null;
  strengths: string | null;

  // Flags
  favorite: boolean;
  save_for_future: boolean;
  contacted: boolean;

  // Pipeline stage (server defaults to "In Consideration")
  stage: CandidateStage;

  // Optional LinkedIn URL (may be stored on ranked row)
  linkedin_url?: string | null;

  // Candidate display fields (prefer these for UI)
  profile_name?: string | null; // from 'search' table or resume person_name
  role?: string | null;
  company?: string | null;

  // --- ADD THIS FIELD ---
  jd_name?: string | null;

  // New field to check if candidate is recommended by current user
  is_recommended?: boolean;

  // Legacy / fallback fields (kept for compatibility across app)
  name?: string | null;
  full_name?: string | null;
  person_name?: string | null;

  // Additional optional fields used elsewhere in the app
  profile_url?: string | null;

  title?: string | null;
  current_title?: string | null;
  current_company?: string | null;
  organization_name?: string | null;
  organization?: string | null;
}

/**
 * ✅ New interface for LinkedIn-sourced candidates (no match_score)
 * Matches the `public.linkedin` table exactly.
 */
export interface LinkedInCandidate {
  linkedin_profile_id: string;
  jd_id: string;
  user_id: string;
  name?: string | null;
  profile_link?: string | null;
  position?: string | null;
  company?: string | null;
  summary?: string | null;
  created_at: string;
  save_for_future?: boolean;
  is_recommended?: boolean;
}